text/XXXX-config_revision.md

```diff
@@ -0,0 +1,268 @@
+
+All that said, using env var configuration should be _optional_, meaning a user employing qri as a library can provide an option to the `lib.NewInstance` constructor function that ignores all environment variables.
+
+### Config Hierarchy
+
+Configuration details should be unified into a single data structure loaded in a predefined hierarchy. From least that hierarchy should be:
+
+* default configuration
+* qri repo configuration file
+* global command-line flags
+
+We can imagine the final computed configuration as a _fold-left_ operation:
+
+```
+config = fold_left(defaultConfig(), ConfigFile(), CliFlags())
+```
+
+Any qri process loads at most 1 configuration file. The source of that file defaults to `$HOME/.qri/config.yaml`, and can be overridden with the `QRI_PATH` environment variable.
+
+In the event that a command line flag and a configuration setting are affecting the same behaviour, the
+command line flag wins, overriding any configuration setting. With that said, **flags should change parameters on lib method calls, not affect configuration**.
+
+
+### Config in alternate contexts:
+
+**qri CLI commands that work without a repo**
+Some commands like `qri setup` have to operate without a `lib.Instance`, and _cannot_ depend on a file, so the loading hierachy simply removes the repoConfig() from the hierarchy, leaving the rest intact:
+
+```
+config = fold_left(defaultConfig(), getEnvVarConfig(), getCLIConfig())
+```
+
+**using qri as a library**
+When using qri as a library, we don't have the `cmd` package, and often supply configuration details directly.
+
+```
+config = fold_left(defaultConfig(), getEnvVarConfig())
+```
+
+### Use multiaddrs, port 0 for random open port
+We should switch all `port` fields to an `address` field instead, and accept a [multiaddr](https://github.com/multiformats/go-multiaddr) string to enhance configurability.
+
+In addition, we should support the convention of using `0` as a port number to indicate "any open port", allowing the operating system to choose.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Remove stale configuration fields
+
+The following fields should be **removed**:
+
+```yaml
+API:
+  proxyforcehttps: false
+  remoteacceptsizemax: 0
+  remoteaccepttimeoutms: 0
+  remotemode: false
+  tls: false
+  urlroot: ""
+P2P:
+  autoNAT: false
+  profilereplication: full
+  bootstrapaddrs: []
+  httpgatewayaddr: ""
+  pubkey: ""
+Repo:
+  middleware: []
+Webapp:
+  analyticstoken: ""
+  enabled: true
+  entrypointhash: /ipfs/QmXofmXcQKnYTMjsfhgTGqzzLLPyS9enaHsfaRR5AxTGBK
+  entrypointupdateaddress: /ipns/webapp.qri.io
+  port: 2505
+Render:
+  defaultTemplateHash: /ipfs/QmeqeRTf2Cvkqdx4xUdWi1nJB2TgCyxmemsL3H4f1eTBaw
+  templateUpdateAddress: /ipns/defaulttmpl.qri.io
+Stats: null
+Update:
+  address: ""
+  daemonize: false
+  type: fs
+```
+
```

Comment:
Assuming this next section of yaml is what we are keeping?
If so, needs an explanation
"The following fields should remain or be added:"